raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2132,15 +2132,32 @@ private long pollLeader(long currentTimeMs) {
             currentTimeMs
         );
 
-        long timeUntilSend = maybeSendRequests(
-            currentTimeMs,
-            partitionState
-                .lastVoterSet()
-                .voterNodes(state.nonAcknowledgingVoters().stream(), channel.listenerName()),
-            this::buildBeginQuorumEpochRequest
-        );
+        long timeUntilNextBeginQuorumSend = state.timeUntilBeginQuorumEpochTimerExpires(currentTimeMs);
+
+        if (timeUntilNextBeginQuorumSend == 0) {
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(channel.listenerName()),
+                this::buildBeginQuorumEpochRequest
+            );
+            state.resetBeginQuorumEpochTimer(currentTimeMs);
+            logger.trace("Attempted to send BeginQuorumEpochRequest as heartbeat to all voters. " +
+                "Request can be retried in {} ms", timeUntilNextBeginQuorumSend);
+        } else if (!state.nonAcknowledgingVoters().isEmpty()) {
```

Comment:
I think that you can remove this `if` case if you set the `beginQuorumEpochTimer` timer to 0 (expired) when becoming leader for the first time.

Reply:
got it, if the non-ack voters miss the beginQuorumEpoch request the first time - we'll rely on the heartbeat to retry requests?